src/snowflake/cli/api/utils/definition_rendering.py

```diff
@@ -305,11 +352,19 @@ def render_definition_template(
             _template_version_warning()
 
         project_definition = build_project_definition(**original_definition)
+        project_context = {CONTEXT_KEY: definition}
         project_context[CONTEXT_KEY]["env"] = environment_overrides
         return ProjectProperties(project_definition, project_context)
 
-    default_env = definition.get("env", {})
-    _validate_env_section(default_env)
+    definition = _add_defaults_to_definition(definition)
+    project_context = {CONTEXT_KEY: definition}
+
+    _validate_env_section(definition.get("env", {}))
+
+    # add available templating functions
+    project_context["id_concat"] = TemplatingFunctions.id_concat
+    project_context["to_id"] = TemplatingFunctions.to_id
+    project_context["to_str"] = TemplatingFunctions.to_str
```

Comment:
to_id and to_str are not currently used - they could be useful, but we can hold off on introducing them yet.